### ISSUE TITLE
[Feature Request] Move-compiler needs to support Callback/Interact function, which is convenient for platform expansion without FS.

### ISSUE THREAD

**Issue by yubing744** [move-language/move#89]
# 🚀 Feature Request

Move-compiler needs to support Callback/Interact function, which is convenient for platform expansion without FS.

## Motivation

I need to implement a Move build in a web browser.

But the browser does not support file operations, compiling move-compiler to wasm-unkown-unkown will fail because it cannot read the file. The solution of the pontem-network team is to let move-compiler provide compilation Callback/Interact, and then wasm implements the file access interface separately. Related Commit: https://github.com/pontem-network/diem/commit/11ae64bb7af73c4f7305fc145d359daede77d19e

But now [move-language/move](https://github.com/move-language/move) has no Callback/Interact function.

## Pitch

**Describe the solution you'd like**
Merge commit: https://github.com/pontem-network/diem/commit/11ae64bb7af73c4f7305fc145d359daede77d19e

**Describe alternatives you've considered**
Rust supports registering custom filesystems

**Are you willing to open a pull request?** (See [CONTRIBUTING](../../CONTRIBUTING.md))
No

## Additional context

My issue:  https://github.com/starcoinorg/starcoin/issues/3363

**Comment by tnowacki** [comment]
@awelc this could be interesting when also looking at the way VS Code integration interacts with virtual files

FWIW having a better sort of different filesystem/virtual file system support would be a great boon to tests too (we wouldn't have to create temp files)

**Comment by yubing744** [comment]
@tnowacki It would be great if virtual filesystems could be supported. What is the PR of this virtual file system, I want to learn about it?

**Comment by awelc** [comment]
> @awelc this could be interesting when also looking at the way VS Code integration interacts with virtual files

Indeed! I think this is the first optimization I would like to focus on, particularly once I am done with current functionality extension for the language server.

> What is the PR of this virtual file system, I want to learn about it?

I don't think there is one at this point. I can start looking at this as, again, I am interested in supporting this, but likely not until next week.

**Comment by yubing744** [comment]
> > @awelc this could be interesting when also looking at the way VS Code integration interacts with virtual files
> 
> Indeed! I think this is the first optimization I would like to focus on, particularly once I am done with current functionality extension for the language server.
> 
> > What is the PR of this virtual file system, I want to learn about it?
> 
> I don't think there is one at this point. I can start looking at this as, again, I am interested in supporting this, but likely not until next week.

Is there progress on this task?

**Comment by awelc** [comment]
I am still working on some foundational support for the IDE. It's still on
the agenda but will have to wait a bit, at least on my side (at least a few
weeks, most likely)

**Comment by yubing744** [comment]
@awelc I wrote an [example](https://github.com/yubing744/wasmer-js-move-compile-demo) of how to compile a move module in the browser using wasmer-js, maybe this is the best solution.
